### ISSUE TITLE
Update for compatibility with pyrate-limiter v3

### ISSUE THREAD

**Issue by JWCook** [JWCook/requests-ratelimiter#78]

**Comment by jamesmhbarry** [comment]
Has there been any progress on this issue? I am getting this error message:

```File "~/anaconda3/lib/python3.11/site-packages/requests_ratelimiter/__init__.py", line 4, in <module>
    from .requests_ratelimiter import *
  File "~/anaconda3/lib/python3.11/site-packages/requests_ratelimiter/requests_ratelimiter.py", line 8, in <module>
    from pyrate_limiter import Duration, Limiter, RequestRate
ImportError: cannot import name 'RequestRate' from 'pyrate_limiter' (~/anaconda3/lib/python3.11/site-packages/pyrate_limiter/__init__.py)```

**Comment by JWCook** [comment]
It's supposed to be pinned to `^2.8`, but it looks like I neglected to do that for the conda-forge recipe. I'll get that updated.

As for moving to pyrate-limiter v3, it seems like there are a few features missing (which the maintainer is still working on), so I'm holding off on this issue until that's sorted out.

**Comment by JWCook** [comment]
This has been fixed on conda-forge in v0.4.2.

**Comment by jamesmhbarry** [comment]
> This has been fixed on conda-forge in v0.4.2.

Great, thanks!

**Comment by desaijimmy** [comment]
> It's supposed to be pinned to `^2.8`, but it looks like I neglected to do that for the conda-forge recipe. I'll get that updated.
> 
> As for moving to pyrate-limiter v3, it seems like there are a few features missing (which the maintainer is still working on), so I'm holding off on this issue until that's sorted out.

Hi JWCook - Thanks for maintaining this lib. Just curious if this can be upgraded to work with pyrate-limiter v3 now? :) 

Thanks!

**Comment by JWCook** [comment]
Probably! I will take another look soon to see if pyrate-limiter v3 has re-implemented all the features I need. If not, I'll submit some issues and/or PRs for whatever is missing.

**Comment by desaijimmy** [comment]
> Probably! I will take another look soon to see if pyrate-limiter v3 has re-implemented all the features I need. If not, I'll submit some issues and/or PRs for whatever is missing.

Look forward to it. Thanks very much!

**Comment by JWCook** [comment]
The main thing still missing from v3 is the `multiprocess`-compatible SQLite backend I originally submitted here: https://github.com/vutran1710/PyrateLimiter/pull/68

That's a pretty important feature that I and others are using via `requests-ratelimiter`. I made an issue for that upstream.

Is there a specific feature you'd like to use in pyrate-limiter v3? Just wondering if there's a short-term workaround I could provide.

**Comment by desaijimmy** [comment]
Thanks for following Jordan - much appreciated!

I've managed to get a workaround for now, so it's not a problem. I just run in to issues when upgrading packages via pip as it results in some conflicts with other packages that I use. However that's a minor issue for now as I can manually sort it out by deleting the latest pyrate-limiter package and just reinstalling requests-ratelimiter, which then installs the appropriate version of pyrate-limiter. 

Hopefully the upstream issues get resolved soon!

Thanks again for following up on this!

**Comment by tgrandje** [comment]
> Is there a specific feature it provides that you'd like to use? Just wondering if there's a short-term workaround I could provide.
@JWCook to add to this discussion, I'm personnally interested by the new PostgresBucket backend.

My use case is a package running API calls but used in different pods/containers: I need to track this whatever the actual number of pods currently running. I was thinking of either trying to create a new custom bucket on a S3 storage system or using the PostgresBucket (which is not supporting async, and I'm not sure about the implications yet)...

_Note : if not that, the package I'm speaking of has it's own way of tracking calls which is stored in a parquet file, which could easily be stored on a S3... But I'm way more in favor of using requests-ratelimiter, though I'm not sure that's (yet) possible..._

Do you have any suggestions?

**Comment by Tyler-Hardin** [comment]
@JWCook, have you checked on this recently? My distro (nixos) switched to pyrate-limiter 3.7 recently and requests-ratelimiter has been marked as broken as a result. Not really your problem, but perhaps a bit of a sign that the world is moving on past <3.0.

**Comment by JWCook** [comment]
Yes, this is still on my radar. I'm now a co-maintainer on pyrate-limiter, and plan to help add/fix the features this library needs. There is quite a [backlog](https://github.com/vutran1710/PyrateLimiter/issues) of other issues to work through, though, so it may be awhile. Any assistance looking into those issues would be appreciated!

My plan is to make a pre-release for requests-ratelimiter (for example, `0.7.0-a0`) with at least minimum support for pyrate-limiter v3 (not necessarily supporting all backends yet), and then make a stable release once the rest of the issues are resolved.

**Comment by JWCook** [comment]
There is a pre-release build available now that's mostly compatible with pyrate-limiter v4:
```sh
pip install requests_ratelimiter==0.9.0.dev0
```

**Comment by JWCook** [comment]
Note: I may need to remove support for `max_delay`, since pyrate-limiter v4 removed this, although it looks like it may be added back in the future. I added a simple busy-wait loop as a placeholder, but I will probably remove it before release.

My own use case for this is an API with some long-term rate limits (per-month limits in addition to per-second/hour); and after hitting the monthly rate limit, I'd rather see an error than delay for the rest of the month. There are some other ways I could handle that, though.

Is anyone else using this feature?
